futaba/cogs/moderation/manual_modaction_warn.py

```diff
@@ -0,0 +1,198 @@
+
+        Returns a list of tuples containing the role and moderator that updated the role on the member.
+        """
+        roles = set(roles)
+
+        updated_roles = []
+
+        async for entry in member.guild.audit_logs(
+            limit=20, action=AuditLogAction.member_role_update
+        ):
+            if entry.target != member:
+                continue
+
+            if entry.user == self.bot.user:
+                continue
+
+            roles_updated_here = roles & (
+                set(entry.before.roles) | set(entry.after.roles)
+            )
+
+            roles -= roles_updated_here
+
+            updated_roles.extend((role, entry.user) for role in roles_updated_here)
+
+        return updated_roles
+
+    async def member_update(self, before, after):
+        member = after
+        after_roles = set(after.roles)
+        before_roles = set(before.roles)
+
+        roles_updated = after_roles ^ before_roles
+
+        if not roles_updated:
+            return
+
+        if not self.bot.sql.settings.get_warn_manual_mod_action(member.guild):
+            return
+
+        special_roles = self.bot.sql.settings.get_special_roles(member.guild)
+        roles_to_check = roles_updated & set(special_roles)
+
+        if not roles_to_check:
+            return
+
+        manually_updated_roles = await self.find_manually_updated_roles(
+            member, roles_to_check
+        )
+
+        special_role_name_action_map = {
+            special_roles.member_role: ManualModActionType.special_role_member,
+            special_roles.guest_role: ManualModActionType.special_role_guest,
+            special_roles.mute_role: ManualModActionType.special_role_mute,
+            special_roles.jail_role: ManualModActionType.special_role_jail,
+        }
+
+        for (role, moderator) in manually_updated_roles:
+            action = special_role_name_action_map[role]
+            await self.dispatch_manual_action_warning(
+                action, moderator, member, role=role
+            )
+
+    async def find_manually_removed_member(self, member):
+        """Finds a possibly manually kicked or banned member.
+
+        Returns a tuple: (remove_type, member, moderator) upon the user being manually banned,
+            otherwise returns None. (where remove_type is either AuditLogAction.kick or AuditLogAction.ban)
+        """
+
+        async for entry in member.guild.audit_logs(limit=20):
+            if entry.action not in {AuditLogAction.ban, AuditLogAction.kick}:
+                continue
+
+            if entry.target != member:
+                continue
+
+            if entry.user == self.bot.user:
+                continue
+
+            return (entry.action, member, entry.user)
+
+    _audit_log_to_manual_mod_action_map = {
+        AuditLogAction.kick: ManualModActionType.kick_member,
+        AuditLogAction.ban: ManualModActionType.ban_member,
+    }
+
+    async def member_remove(self, member):
```

Comment:
The ban event is still fired after leave events afaik.

That util function would be good, is it alright to make it a static method so that I can use it without having to get the instance of the tracker cog?

Reply:
Right, tracker cog isn't mandatory. What do you mean by making it static though, where would it go? Alternatively, you could make the tracker cog mandatory, and then add a link to it in `client.py` (see journal). Not sure if we want to do that though.